contracts/contracts/coordination/subscription/BqETHSubscription.sol

```diff
@@ -0,0 +1,261 @@
+     * @param additionalEncryptorSlots Additional number of slots for encryptors
+     */
+    function payForEncryptorSlots(uint128 additionalEncryptorSlots) external {
+        uint256 currentPeriodNumber = getCurrentPeriodNumber();
+        Billing storage billing = billingInfo[currentPeriodNumber];
+        require(billing.paid, "Current billing period must be paid");
+
+        uint32 duration = subscriptionPeriodDuration;
+        uint32 endOfCurrentPeriod = 0;
+        if (startOfSubscription != 0) {
+            endOfCurrentPeriod = uint32(
+                startOfSubscription + (currentPeriodNumber + 1) * subscriptionPeriodDuration
+            );
+            duration = endOfCurrentPeriod - uint32(block.timestamp);
+        }
+
+        uint256 fees = encryptorFees(additionalEncryptorSlots, duration);
+        billing.encryptorSlots += additionalEncryptorSlots;
+
+        feeToken.safeTransferFrom(msg.sender, address(this), fees);
+        emit EncryptorSlotsPaid(msg.sender, fees, additionalEncryptorSlots, endOfCurrentPeriod);
+    }
+
+    /**
+     * @notice Withdraws the contract balance to the treasury
+     * @param amount The amount to withdraw
+     */
+    function withdrawToTreasury(uint256 amount) external onlyOwner {
+        require(amount <= feeToken.balanceOf(address(this)), "Insufficient balance available");
+        feeToken.safeTransfer(msg.sender, amount);
+        emit WithdrawalToTreasury(msg.sender, amount);
+    }
+
+    function processRitualPayment(
+        address initiator,
+        uint32 ritualId,
+        uint256 numberOfProviders,
+        uint32 duration
+    ) external override onlyCoordinator {
+        require(initiator == adopter, "Only adopter can initiate ritual");
+        if (startOfSubscription == 0) {
+            startOfSubscription = uint32(block.timestamp);
+        }
+        uint32 endOfSubscription = getEndOfSubscription();
+        require(endOfSubscription != 0, "Subscription has to be paid first");
+        require(
+            endOfSubscription + yellowPeriodDuration + redPeriodDuration >=
+                block.timestamp + duration &&
+                numberOfProviders <= maxNodes,
+            "Ritual parameters exceed available in package"
+        );
+        require(
+            address(accessController) != address(0) &&
+                accessController == coordinator.getAccessController(ritualId),
+            "Access controller for ritual must be approved"
+        );
+
+        if (activeRitualId != INACTIVE_RITUAL_ID) {
+            Coordinator.RitualState state = coordinator.getRitualState(activeRitualId);
+            require(
+                state == Coordinator.RitualState.DKG_INVALID ||
+                    state == Coordinator.RitualState.DKG_TIMEOUT ||
+                    state == Coordinator.RitualState.EXPIRED, // TODO check if it's ok
+                "Only failed rituals allowed to be reinitiated"
```

Comment:
Since "expired" rituals aren't really "failed":
```suggestion
                "Only failed/expired rituals allowed to be reinitiated"
```

We could have used "non-active" but there is a specific `RitualState.ACTIVE` which gives "non-active" a different meaning than intended here.